File: leetcode/go/1987.number-of-unique-good-subsequences.go
func numberOfUniqueGoodSubsequences(binary string) int {
    dp0, dp1 := 0, 0
    zero := 0
    for i:=0; i < len(binary); i++ {
        if binary[i] == '0' {
            dp0 = (dp0 + dp1) % 1_000_000_007
            zero = 1
        } else {
            dp1 = (dp1 + dp0 + 1) % 1_000_000_007
        }
    }
    return (dp0 + dp1 + zero) % 1_000_000_007
}

File: leetcode/rust-leetcode/src/1987.number-of-unique-good-subsequences.rs
static k_mod: usize = 1_000_000_007;
impl Solution {
    pub fn number_of_unique_good_subsequences(binary: String) -> i32 {
        let bi = binary.into_bytes().into_iter().map(|b| (b-b'0') as usize).collect::<Vec<_>>();
        let mut dp = [0, 0];
        let mut zero = false;
        for &b in bi.iter() {
            dp[b] = (dp[0] + dp[1]) % k_mod;
            if b == 1 {
                dp[1] += 1;
            } else {
                zero = true;
            }
        }
        // add 1 for '0' if there is zero in the binary string
        ((dp[0]+dp[1] + if zero {1} else {0}) % k_mod) as _
    }
}

